ruma-push-gateway-api/src/push/send_event_notification/v1.rs

```diff
@@ -0,0 +1,371 @@
+
+    /// The `content` field from the event, if present. The pusher may omit this
+    /// if the event had no content or for any other reason.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub content: Option<Box<RawJsonValue>>,
+
+    /// This is a dictionary of the current number of unacknowledged
+    /// communications for the recipient user. Counts whose value is zero should
+    /// be omitted.
+    #[serde(default, skip_serializing_if = "ruma_serde::is_default")]
+    pub counts: NotificationCounts,
+
+    /// This is an array of devices that the notification should be sent to.
+    pub devices: &'a [Device],
+}
+
+impl<'a> Notification<'a> {
+    /// Create a new notification for the given devices
+    pub fn new(devices: &'a [Device]) -> Self {
+        Notification { devices, ..Default::default() }
+    }
+}
+
+/// Type for passing information about notification priority.
+///
+/// This may be used by push gateways to deliver less time-sensitive
+/// notifications in a way that will preserve battery power on mobile devices.
+#[derive(Clone, Copy, Debug, Serialize, Deserialize, Display, EnumString, PartialEq)]
+#[serde(rename_all = "snake_case")]
+#[strum(serialize_all = "snake_case")]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub enum NotificationPriority {
+    /// A high priority notification
+    High,
+    /// A low priority notification
+    Low,
+}
+
+impl Default for NotificationPriority {
+    fn default() -> Self {
+        Self::High
+    }
+}
+
+/// Type for passing information about notification counts.
+#[derive(Clone, Copy, Debug, Default, Deserialize, Serialize, PartialEq)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub struct NotificationCounts {
+    /// The number of unread messages a user has across all of the rooms they
+    /// are a member of.
+    #[serde(default, skip_serializing_if = "ruma_serde::is_default")]
+    pub unread: UInt,
+
+    /// The number of unacknowledged missed calls a user has across all rooms of
+    /// which they are a member.
+    #[serde(default, skip_serializing_if = "ruma_serde::is_default")]
+    pub missed_calls: UInt,
+}
+
+impl NotificationCounts {
+    /// Create new notification counts from the given unread and missed call
+    /// counts.
+    pub fn new(unread: UInt, missed_calls: UInt) -> Self {
+        NotificationCounts { unread, missed_calls }
+    }
+}
+
+/// Type for passing information about devices.
+#[derive(Clone, Debug, Deserialize, Outgoing, Serialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub struct Device {
+    /// The `app_id` given when the pusher was created.
+    ///
+    /// Max length, 64 chars.
+    pub app_id: String,
+
+    /// The `pushkey` given when the pusher was created.
+    ///
+    /// Max length, 512 bytes.
+    pub pushkey: String,
+
+    /// The unix timestamp (in seconds) when the pushkey was last updated.
```

Comment:
If this timestamp is really in seconds, the `ruma_serde::time::opt_ms_since_unix_epoch` module is wrong and we need an `opt_s_since_unix_epoch` module additionally.

Reply:
Good catch, added the serde module as the spec stats seconds and synapse uses seconds https://github.com/matrix-org/synapse/blob/c9dbee50aefc22390f600a0219ca7fa1ae9acd88/synapse/push/httppusher.py#L324